sql/src/main/java/io/crate/planner/operators/TopNDistinct.java

```diff
@@ -0,0 +1,125 @@
+import io.crate.execution.engine.pipeline.TopN;
+import io.crate.expression.symbol.InputColumn;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.RowGranularity;
+import io.crate.planner.ExecutionPlan;
+import io.crate.planner.Merge;
+import io.crate.planner.PlannerContext;
+import io.crate.types.DataTypes;
+
+public final class TopNDistinct extends ForwardingLogicalPlan {
+
+    private final Symbol limit;
+    private final List<Symbol> outputs;
+
+    public TopNDistinct(LogicalPlan source, Symbol limit, List<Symbol> outputs) {
+        super(source);
+        this.limit = limit;
+        this.outputs = outputs;
+    }
+
+    public Symbol limit() {
+        return this.limit;
+    }
+
+    @Override
+    public ExecutionPlan build(PlannerContext plannerContext,
+                               ProjectionBuilder projectionBuilder,
+                               int limitHint,
+                               int offsetHint,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               SubQueryResults subQueryResults) {
+        var executionPlan = source.build(
+            plannerContext,
+            projectionBuilder,
+            TopN.NO_LIMIT,
+            TopN.NO_OFFSET,
+            null,
+            null,
+            params,
+            subQueryResults
+        );
+        if (executionPlan.resultDescription().hasRemainingLimitOrOffset()) {
+            executionPlan = Merge.ensureOnHandler(executionPlan, plannerContext);
+        }
+        assert source.outputs().equals(outputs)
+            : "source outputs must match outputs, TopNDistinct can't re-evaluate outputs";
+        int limit = DataTypes.INTEGER.value(
+            evaluate(
+                plannerContext.transactionContext(),
+                plannerContext.functions(),
+                this.limit,
+                params,
+                subQueryResults
+            )
+        );
+        var inputColOutputs = InputColumn.mapToInputColumns(outputs);
+        executionPlan.addProjection(
+            new TopNDistinctProjection(
+                limit,
+                inputColOutputs,
+                source.preferShardProjections() ? RowGranularity.SHARD : RowGranularity.CLUSTER
+            )
+        );
+        boolean onHandler = ExecutionPhases.executesOnHandler(
+            plannerContext.handlerNode(), executionPlan.resultDescription().nodeIds());
+        if (!onHandler) {
+            executionPlan = Merge.ensureOnHandler(executionPlan, plannerContext);
+            executionPlan.addProjection(
+                new TopNDistinctProjection(limit, inputColOutputs, RowGranularity.CLUSTER));
+        } else if (source.preferShardProjections()) {
+            executionPlan.addProjection(
+                new TopNDistinctProjection(limit, inputColOutputs, RowGranularity.CLUSTER));
```

Comment:
shouldn't this run on a shard by `RowGranularity.SHARD`?

Reply:
The shard level projection is added here: https://github.com/crate/crate/pull/9371/files/baf1f93e9fb9c8fbf259a25b77f7b3b1bf09b03e#diff-8c85ee8c0bf4a65ff58f4429ab707d3bR99

The `CLUSTER` projection here is if we're already on the handler and used a shard projection -> We need to merge the N*numShards to N.

Reply:
ah right. I think the code is fine if read carefully ;-)